src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java

```diff
@@ -1396,28 +1367,38 @@ public Embargo findEmbargo(Long id) {
         return d.getEmbargo();
     }
     
-    public Long getStorageUsageByCreator(AuthenticatedUser user) {
-        Query query = em.createQuery("SELECT SUM(o.filesize) FROM DataFile o WHERE o.creator.id=:creatorId");
-        
-        try {
-            Long totalSize = (Long)query.setParameter("creatorId", user.getId()).getSingleResult();
-            logger.info("total size for user: "+totalSize);
-            return totalSize == null ? 0L : totalSize; 
-        } catch (NoResultException nre) { // ?
-            logger.info("NoResultException, returning 0L");
-            return 0L;
+    /**
+     * Checks if the supplied DvObjectContainer (Dataset or Collection; although
+     * only collection-level storage quotas are officially supported as of now)
+     * has a quota configured, and if not, keeps checking if any of the direct
+     * ancestor Collections further up have a configured quota. If it finds one, 
+     * it will retrieve the current total content size for that specific ancestor 
+     * dvObjectContainer and use it to define the quota limit for the upload
+     * session in progress. 
+     * 
+     * @param parent - DvObjectContainer, Dataset or Collection
+     * @return upload session size limit spec, or null if quota not defined on 
+     * any of the ancestor DvObjectContainers
+     */
+    public UploadSessionQuotaLimit getUploadSessionQuotaLimit(DvObjectContainer parent) {
```

Comment:
is it possible to define a quote of say 100 GB for a collection, and a larger quote for a child collection (say 1 TB)? in this case, does the child size effectively override the parent?

Reply:
Yes, I should have mentioned this in the PR, when describing my implementation (will add): the quota defined on the nearest ancestor applies, if any, in all cases. 
Answering your specific question, yes, it is possible to give a child collection a quota larger than the parent's, which it will override. 

But the way my implementation works, once the child collection goes over the quota limit defined for the parent,  that will set put the parent's storage use over their limit, so nobody will be able to add any content to the datasets immediately under the parent, or in any other subcollections thereof that don't have their own quota defined. So I should be telling site admins not to do this in the guide. 

(Is this a problem?)

Reply:
... Maybe not telling them "not to do this", but rather, spell it out that you only want to do this if you want to specifically make a collection read-only/closed for more uploads, while keeping one or more of the sub-collections open for business - ?

Reply:
> so does this replace what was there then for (not used,, I know) user quotas - I saw several places where there seemed to be some straghtforward replacing and figure it;s just easoer to ask here.

It uses/builds on top of the mechanics of quota enforcement added in that phase one PR. The object that was called "UserStorageQuota" is still being used, but it was renamed to UploadSessionQuota... to better reflect what it does (it was moved out of DataFileServiceBean also). The only way to actually define a quota in that phase one pr was to set an instance-wide limit, which then applied to all users; AND "applied" in a sense of counting against the sum of filesizes by `dvobject.creator_id`. A limit defined like that had no practical value obviously; it was only used to test the mechanisms of enforcement/the UI etc. It was removed from this branch. 
Does this make sense?

Reply:
I should add that, if we want to add real per-user quotas going forward, I even left a commented out relation to AuthenticatedUser in the new StorageQuota class. So, the same entity can be used for that as well (but we would need to figure out how/what to count as storage use by specific users). There's also a comment saying that nothing should be stopping us from changing the DvObject relation there to `@ManyToOne`, and then be able to define quotas on specific collections/datasets AND for specific users... but I don't know if we ever want to go there.